### PR TITLE
#25573 add autosave on templates portlet

### LINKED ISSUE #25573: Template Builder: Implement Autosave in the template portlet
### User Story
As a dotCMS user, I want to have the ability to autosave my work in the template builder, so I can more easily track changes and avoid data loss.

### Acceptance Criteria
- Autosave should be enabled by default
- Autosave should save the template every 5 seconds
- Autosave should have the same approach as the layout editor

### Proposed Objective
Core Feature

### Proposed Priority
Priority 3 - Average

### External Links... Slack Conversations, Support Tickets, Figma Designs, etc.:
N/A

### Assumptions & Initiation Needs
- Autosave should not interfere with any existing functionality 
- Autosave should work across all browsers

### Quality Assurance Notes & Workarounds
N/A

### Sub-Tasks & Estimates:
N/A

### PR DESCRIPTION
**Pull request by danielmdob**
### <samp>🤖 Generated by Copilot at 9dd985b</samp>

This pull request refactors and simplifies the dot-edit-layout and dot-template-builder features to use the new template builder library component and the DotRouterService. It also adds guard services to prevent losing unsaved changes when navigating away from the layout editing routes. It updates and removes tests and styles accordingly. It affects the following files: `dot-edit-layout.service.spec.ts`, `dot-edit-layout.service.ts`, `dot-router.service.ts`, `can-deactivate-guard.service.ts`, `dot-edit-page-routing.module.ts`, `dot-edit-layout.component.html`, `dot-edit-layout.component.scss`, `dot-edit-layout.component.spec.ts`, `dot-edit-layout.component.ts`, `dot-template-builder.component.html`, `dot-template-builder.component.spec.ts`, `dot-template-builder.component.ts`, `dot-template-create-edit-routing.module.ts`, `dot-template-new-routing.module.ts`, `dot-template-new.module.ts`, `dot-template.store.spec.ts`, `layout-editor-can-deactivate-guard.service.spec.ts`, `dot-router.service.spec.ts`, and `can-deactivate-guard.service.spec.ts`.

### Checklist
- [X] Tests
- [ ] Translations
- [ ] Security Implications Contemplated (add notes if applicable)

### Additional Info
** any additional useful context or info **

### Screenshots
Original             |  Updated
:-------------------------:|:-------------------------:
image  |  image

### CODE REVIEW

File: core-web/apps/dotcms-ui/src/app/api/services/dot-router/dot-router.service.spec.ts [hunk 1]
```diff
@@ -246,6 +246,21 @@ describe('DotRouterService', () => {
...
+            done();
+        });
+    });
+
+    it('trigger pageLeaveRequest observable stream when a page leave is requested', (done) => {
+        service.pageLeaveRequest$.subscribe(() => {
+            done();
+        });
+        service.requestPageLeave();
+    });
```
Comment (zJaaal) [inline core-web/apps/dotcms-ui/src/app/api/services/dot-router/dot-router.service.spec.ts:262]: should we add an spy instead of empty subscribe?

**Review by zJaaal — APPROVED** [review]

File: core-web/apps/dotcms-ui/src/app/portlets/dot-templates/dot-template-create-edit/dot-template-builder/dot-template-builder.component.ts [hunk 2]
```diff
@@ -47,5 +73,25 @@ export class DotTemplateBuilderComponent implements OnInit {
...
             this.historyIframe.iframeElement.nativeElement.contentWindow.location.reload();
         }
+
+        this.dotRouterService.forbidRouteDeactivation();
+        this.lastTemplate = item;
+
+        this.templateUpdate$.next(item);
+    }
+
+    saveTemplateDebounce() {
```
Comment (fmontes) [inline core-web/apps/dotcms-ui/src/app/portlets/dot-templates/dot-template-create-edit/dot-template-builder/dot-template-builder.component.ts:83]:
```suggestion
    private saveTemplateDebounce() {
```

File: core-web/libs/template-builder/src/lib/components/template-builder/template-builder.component.spec.ts [hunk 3]
```diff
@@ -33,14 +33,16 @@ global.structuredClone = jest.fn((val) => {
 });
 
 describe('TemplateBuilderComponent', () => {
-    let spectator: SpectatorHost<TemplateBuilderComponent>;
+    // let spectator: SpectatorHost<TemplateBuilderComponent>;
```
Comment (fmontes) [inline core-web/libs/template-builder/src/lib/components/template-builder/template-builder.component.spec.ts:36]: Clean up

File: core-web/libs/template-builder/src/lib/components/template-builder/template-builder.component.spec.ts [hunk 4]
```diff
@@ -119,27 +120,35 @@ describe('TemplateBuilderComponent', () => {
         const totalBoxes = FULL_DATA_MOCK.rows.reduce((acc, row) => {
             return acc + row.columns.length;
         }, 0);
-
-        expect(spectator.queryAll(byTestId('box')).length).toBe(totalBoxes);
+        expect(spectator.queryAll(byTestId(/builder-box-\d+/g)).length).toBe(totalBoxes);
```
Comment (fmontes) [inline core-web/libs/template-builder/src/lib/components/template-builder/template-builder.component.spec.ts:123]: Is still 12?

File: core-web/libs/template-builder/src/lib/components/template-builder/template-builder.component.spec.ts [hunk 5]
```diff
@@ -219,7 +228,7 @@ describe('TemplateBuilderComponent', () => {
 
             spectator.detectChanges();
 
-            store.init({
+            store.setState({
```
Comment (fmontes) [inline core-web/libs/template-builder/src/lib/components/template-builder/template-builder.component.spec.ts:231]: Are we sure we need this whole tests here? Looks like is testing the store only.

**Review by fmontes — CHANGES_REQUESTED** [review]

File: core-web/libs/template-builder/src/lib/components/template-builder/template-builder.component.spec.ts (hunk 4 above)
Comment (danielmdob) [reply core-web/libs/template-builder/src/lib/components/template-builder/template-builder.component.spec.ts:123]: Yes

**Review by fmontes — APPROVED** [review]

**Pull request merged**